bundles/org.openhab.binding.clementineremote/src/main/java/org/openhab/binding/clementineremote/internal/ClementineRemoteHandler.java

```diff
@@ -0,0 +1,414 @@
+        }
+    }
+
+    private void handleStop() {
+        setTrack("-", "-", "-", "-");
+        setState(State.stopped);
+        handleNullableTrackPos(null);
+    }
+
+    private void handleNullableTrackPos(@Nullable Integer newPos) {
+        if (newPos != null) {
+            setState(State.playing);
+            currentPos = newPos;
+        } else {
+            currentPos = 0;
+        }
+
+        updateState(CHANNEL_POSITION, new QuantityType<>(currentPos, Units.SECOND));
+    }
+
+    private void handleTrackPos(ResponseUpdateTrackPosition message) {
+        handleNullableTrackPos(message.getPosition());
+    }
+
+    private void handleVolume(RequestSetVolume message) {
+        volume = message.getVolume();
+        updateState(CHANNEL_VOLUME, new PercentType(volume));
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(ClementineRemoteConfiguration.class);
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        scheduler.execute(this::connect);
+    }
+
+    private void sleep15secs() {
+        try {
+            sleep(15000);
+        } catch (InterruptedException e) {
+            logger.warn("sleep interrupted: {}", e.getMessage());
+        }
+    }
+
+    private void sendConnectMessage() throws IOException {
+        var req = builder.getRequestConnectBuilder();
+        if (config.authCode != null && config.authCode != 0) {
+            req.setAuthCode(config.authCode);
+        }
+        sendMessageOrThrow(builder.setRequestConnect(req.build()).build());
+    }
+
+    private boolean sendSkip(int d) {
+        state = State.skipping;
+        var pos = RequestSetTrackPosition.newBuilder().setPosition(currentPos + d);
+        Message msg = builder.setType(MsgType.SET_TRACK_POSITION).setRequestSetTrackPosition(pos).build();
+        return sendMessage(msg);
+    }
+
+    private boolean sendMessage(MsgType type) {
+        state = State.unknown;
+        return sendMessage(builder.setType(type).build());
+    }
+
+    private boolean sendMessage(Message message) {
+        if (out == null) {
+            return false;
+        }
+        try {
+            sendMessageOrThrow(message);
+            return true;
+        } catch (IOException e) {
+            logger.warn("sendMessage({}) failed: {}", message, e.getMessage());
+            disconnect();
+            return false;
+        }
+    }
+
+    private void sendMessageOrThrow(Message message) throws IOException {
+        byte[] bytes = message.toByteArray();
+        out.writeInt(bytes.length);
+        out.write(bytes);
+        out.flush();
+    }
+
+    private boolean sendVolume(int percent) {
+        var volume = RequestSetVolume.newBuilder().setVolume(percent);
+        Message msg = builder.setType(MsgType.SET_VOLUME).setRequestSetVolume(volume).build();
+        return sendMessage(msg);
+    }
+
+    private boolean setState(State newState) {
```

Comment:
Not wrong, but for readability, consider rename to `updatePlayerStateChannels`

Reply:
fixed.

Reply:
I meant to rename the `setState` to `updatePlayerStateChannels` and keep the switch in the method. But for you consideration.

Reply:
As written before, I refactored the whole channel update mechanism.